### PR TITLE
Set sky API: Add bool for clouds in front of custom skybox 

### PR DESCRIPTION
**Pull request by paramat**
Default true.
Add 'm_clouds_enabled' bool to sky.h, set from new bool in 'set sky' API.
Make 'getCloudsVisible()' depend on 'm_clouds_enabled' instead of
'm_visible' (whether normal sky is visible).
///////////////////

screenshot_20170428_024946

PR for issue #5669

### CODE REVIEW

**Comment by sofar** [comment]
just disable the clouds if the cloud density is `0.0`, then you don't need to add yet another bool.

clouds should be possible in all skybox types, not just regular. hence the arbitrary removal in my PR.

**Comment by sofar** [comment]
Yeah, this PR **kills** the one thing I wanted: clouds with skybox textures. :(

```
diff --git a/src/game.cpp b/src/game.cpp
index 32d229fa..24fba150 100644
--- a/src/game.cpp
+++ b/src/game.cpp
@@ -3247,9 +3247,8 @@ void Game::processClientEvents(CameraOrientation *cam)
 
                case CE_SET_SKY:
                        sky->setVisible(false);
-                       // Whether clouds are visible in front of a custom skybox
-                       // To be set by a new set_sky bool parameter (default false)
-                       sky->setCloudsEnabled(false);
+                       // paramat can replace this with some code that checks cloud density == 0.0f
+                       sky->setCloudsEnabled(true);
 
                        if (skybox) {
                                skybox->remove();
```

**Comment by sofar** [comment]
@paramat, with the cloud API, wouldn't you logically place the `clouds on/off` bool in the cloud API?

just seems the logical place.

**Comment by nerzhul** [comment]
setCloudEnabled -> enableClouds

**Comment by paramat** [comment]
> clouds should be possible in all skybox types, not just regular

That's what this PR will do, it's WIP.

> Yeah, this PR kills the one thing I wanted: clouds with skybox textures.

That feature is what i want too and is the reason for the PR, it's WIP:

> First step tested and works, the setCloudsEnabled() bool set in SET_SKY decides whether clouds are visible in front of a custom skybox.
Next, add a 'clouds' bool parameter to the set_sky API.

Please read #5669 too.

**Comment by paramat** [comment]
> setCloudEnabled -> enableClouds

setCloudsEnabled is chosen to be consistent with setVisible, it's 'set' followed by the variable name.

**Comment by paramat** [comment]
> just disable the clouds if the cloud density is 0.0

Another problem is, if the showing of clouds in front of a skybox is automatic according to cloud density it will be 'on' by default, which will break mods, my suggested setting will be false by default to not break mods.

**Comment by sofar** [comment]
Well finish your PR already, as long as it can be enabled I'm fine with it since it's all new APIs anyway.

:P

**Comment by paramat** [comment]
Updated but have this compiler error i cannot yet understand:
```
/home//minetestf/gitmods/minetest/src/game.cpp: In member function ‘void Game::processClientEvents(CameraOrientation*)’:
/home//minetestf/gitmods/minetest/src/game.cpp:3259:41: error: invalid type argument of unary ‘*’ (have ‘bool’)
    sky->setCloudsEnabled(*event.set_sky.clouds);
                                         ^
/home//minetestf/gitmods/minetest/src/game.cpp:3293:25: error: type ‘bool’ argument given to ‘delete’, expected pointer
    delete event.set_sky.clouds;
                         ^
```
Also nto sure about my packet order here, would it better to have the new bool added in the middle somewhere instead of at the end?
```
void Server::SendSetSky(u16 peer_id, const video::SColor &bgcolor,
		const std::string &type, const std::vector<std::string> &params,
		bool &clouds)
{
	NetworkPacket pkt(TOCLIENT_SET_SKY, 0, peer_id);
	pkt << bgcolor << type << (u16) params.size();

	for(size_t i=0; i<params.size(); i++)
		pkt << params[i];

	pkt << clouds;

	Send(&pkt);
}
```

File: src/game.cpp [hunk 1]
```diff
@@ -3287,6 +3290,7 @@ void Game::processClientEvents(CameraOrientation *cam)
 			delete event.set_sky.bgcolor;
 			delete event.set_sky.type;
 			delete event.set_sky.params;
+			delete event.set_sky.clouds;
```
Comment (sofar) [inline src/game.cpp]: You don't delete `bool` members. Only `class` members need deletion.

File: src/game.cpp [hunk 2]
```diff
@@ -3255,6 +3255,8 @@ void Game::processClientEvents(CameraOrientation *cam)
 
 		case CE_SET_SKY:
 			sky->setVisible(false);
+			// Whether clouds are visible in front of a custom skybox
+			sky->setCloudsEnabled(*event.set_sky.clouds);
```
Comment (sofar) [inline src/game.cpp]: just remove the `*`

**Comment by sofar** [comment]
And yes, new members go at the end.

**Comment by sofar** [comment]
If I fix up the two issues and compile, it runs but immediately crashes on this line in `l_object.cpp`

```
while (lua_next(L, table) != 0) {
```

I suspect it is somehow because the extra member in the args is mucking up the parsing algorithm. Lua knowledge required to fix.

**Comment by sofar** [comment]
Ahh of course, @paramat, apply this `diff` to fix the code and make it work:

```
diff --git a/src/game.cpp b/src/game.cpp
index 5ab70bae..17ac49a0 100644
--- a/src/game.cpp
+++ b/src/game.cpp
@@ -3248,7 +3248,7 @@ void Game::processClientEvents(CameraOrientation *cam)
 		case CE_SET_SKY:
 			sky->setVisible(false);
 			// Whether clouds are visible in front of a custom skybox
-			sky->setCloudsEnabled(*event.set_sky.clouds);
+			sky->setCloudsEnabled(event.set_sky.clouds);
 
 			if (skybox) {
 				skybox->remove();
@@ -3282,7 +3282,6 @@ void Game::processClientEvents(CameraOrientation *cam)
 			delete event.set_sky.bgcolor;
 			delete event.set_sky.type;
 			delete event.set_sky.params;
-			delete event.set_sky.clouds;
 			break;
 
 		case CE_OVERRIDE_DAY_NIGHT_RATIO:
diff --git a/src/script/lua_api/l_object.cpp b/src/script/lua_api/l_object.cpp
index be64836b..c24c0c39 100644
--- a/src/script/lua_api/l_object.cpp
+++ b/src/script/lua_api/l_object.cpp
@@ -1676,9 +1676,8 @@ int ObjectRef::l_set_sky(lua_State *L)
 
 	std::vector<std::string> params;
 	if (lua_istable(L, 4)) {
-		int table = lua_gettop(L);
 		lua_pushnil(L);
-		while (lua_next(L, table) != 0) {
+		while (lua_next(L, 4) != 0) {
 			// key at index -2 and value at index -1
 			if (lua_isstring(L, -1))
 				params.push_back(lua_tostring(L, -1));
```

**Comment by sofar** [comment]
With the extra changes in my patch, I was able to make the `skybox` mod work and keep clouds enabled.

This was the `diff` to the `skybox` mod that I used:

```
diff --git a/init.lua b/init.lua
index c5ea4a2..d5c0cec 100644
--- a/init.lua
+++ b/init.lua
@@ -39,7 +39,7 @@ skybox.set = function(player, number)
                sky[1] .. "Back.jpg",
                sky[1] .. "Left.jpg",
                sky[1] .. "Right.jpg",
-       })
+       }, true)
        player:set_attribute("skybox:skybox", sky[1])
 end
 
@@ -49,7 +49,7 @@ end
 
 skybox.clear = function(player)
        player:override_day_night_ratio(nil)
-       player:set_sky("white", "regular")
+       player:set_sky("white", "regular", nil, true)
        player:set_attribute("skybox:skybox", "off")
 end
 
```

**Comment by paramat** [comment]
Thanks will apply.

**Comment by rubenwardy** [comment]
:-1:

this should be done using the clouds API

**Comment by paramat** [comment]
> what the intended parameter does is control 'show clouds in front of custom skybox' so should be in the custom skybox API.

To do it in the clouds API would mean a new bool 'show clouds with custom skybox', density cannot be used as explained above.
So the new bool in the clouds API would be one only relevant to custom skybox use, so it would be out of place there, and in the right place in the set sky API.
The bool is useless when only using the clouds API, as it only has effect when a custom skybox is used, and when using a custom skybox the bool is very often needed. So this would mean unnecessary use of 2 APIs when using custom skyboxes, with no benefit to having the bool in the clouds API.

**Comment by rubenwardy** [comment]
I think it is to enable clouds by default with skybox, it's a small nongame breaking breakage that it easy to fix by mods, and results in a nicer api

**Comment by paramat** [comment]
Sofar,
`player:set_sky("white", "regular", nil, true)`
To reset sky to normal you don't need the 'true' because:
```
			if (*event.set_sky.type == "regular") {
				sky->setVisible(true);
				sky->setCloudsEnabled(true);
```

**Comment by paramat** [comment]
Rubenwardy we could consider making the default of the bool true (with the bool in set sky), but all users of custom skybox are using it without clouds (obviously i know) and in future i expect it will be a 50%/50% split who wants clouds with a skybox. So i still think default 'false' makes slightly more sense.

**Comment by paramat** [comment]
Ok i don't feel strongly about the default so since you both want that i'll make it default true.
I'll sk c55 to check this since 'set sky' is his code.

**Comment by sofar** [comment]
I actually don't mind it being as part of the sky, we could see for instance a separate API for sun/moon etc, dangling underneath the sky API.

**Comment by paramat** [comment]
Updated to default true, some code improved:
```
+	bool clouds = true;
+	if (lua_isboolean(L, 5))
+		clouds = lua_toboolean(L, 5);
```

**Comment by paramat** [comment]
Tested and fine, will ask c55 to check the code tomorrow.

**Comment by sofar** [comment]
>`player:set_sky("white", "regular", nil, true)`
>To reset sky to normal you don't need the 'true' because:
>```
>			if (*event.set_sky.type == "regular") {
>				sky->setVisible(true);
>				sky->setCloudsEnabled(true);
>```

I know, but the original code you posted crashed here and therefore I left it to make sure the crash was fixed.

**Comment by nerzhul** [comment]
what is this ugly bar on the screen ? don't tell me it's rain.

**Comment by rubenwardy** [comment]
Now it's on by default, I don't object - however surely it's a better and simpler API to require this?

```lua
player:set_sky({ skybox = "bleh" })
player:set_clouds({ density = 0 })
```

rather than

```lua
player:set_sky({ skybox = "bleh" }, true)
```

**Comment by paramat** [comment]
There are multiple ugly bars, not part of the comimit, that's a mod.

Using 2 APIs is not better or simpler as explained.

**Comment by sofar** [comment]
@paramat please rebase, I'd like to test and get this merged (and use the changes in the `skybox` mod).

**Comment by paramat** [comment]
Will do today.

**Comment by paramat** [comment]
Rebased.

**Comment by paramat** [comment]
Tested.

**Comment by paramat** [comment]
Improved commit message, now ready for merge.

**Comment by sofar** [comment]
:+1: looks great, and I love that it's by default enabled.

**Comment by paramat** [comment]
Lua_api.txt updated. Please merge tonight if it tests ok for you.

File: src/sky.h [hunk 3]
```diff
@@ -65,10 +65,12 @@ class Sky : public scene::ISceneNode
 		return m_visible ? m_skycolor : m_fallback_bg_color;
 	}
 
-	bool getCloudsVisible() { return m_clouds_visible && m_visible; }
+	bool getCloudsVisible() { return m_clouds_visible && m_clouds_enabled; }
```
Comment (nerzhul) [inline src/sky.h]: missing constness !

**Pull request merged**